File: X2D.py
import construct.core
import io

from construct import *
from enum import IntEnum


class Device(IntEnum):
    Tyxia_ZZAA = 0
    Calybox = 3
    Deltia_Emitter = 5
    Tydom_Panel_Controller = 13
    Tyxia_XXYY = 18
    Two_Buttons_Alarm_Remote = 21
    Four_Buttons_Alarm_Remote = 22

    # Detector/Simple
    Volumetric = 32
    Volumetric_Infrared = 33
    Volumetric_Pressure = 34
    Volumetric_Dual = 35
    Perimetric = 40
    Perimetric_Contact = 41
    Perimetric_Glass_Breakage = 42
    Perimetric_Sound = 43
    Perimetric_Roller_Shutter = 44
    Technical = 48
    Technical_Water = 49
    Technical_Gas = 50
    Technical_Fire = 51
    Technical_Smoke = 52
    Technical_Frost = 53
    Technical_Power_Outage = 54
    Technical_Phone_Outage = 55
    Technical_Freezer_Outage = 56

    USB_Key = 62


class Family(IntEnum):
    Regulation = 1
    Sensor = 2
    Metering = 3
    Actuator = 4


Affiliations = {
    1: Family.Regulation,
    2: Family.Regulation,
    10: Family.Sensor,
    12: Family.Sensor,
    14: Family.Metering,
    33: Family.Actuator,
    34: Family.Actuator,
    35: Family.Actuator,
    37: Family.Actuator,
    252: Family.Metering
}


class Attribute(IntEnum):
    Simple = 1
    WithData = 5


#
# WithData
#

class MessageDataType(IntEnum):
    Enrollment = 0  # g EnrolmentMessage
    HeatingLevel = 1  # m regulation Confort heating level
    FunctioningLevel = 2  # n regulation Auto functioning mode
    InternalTemperature = 10  # p Internal temperature command
    ExternalTemperature = 11  # o External temperature command
    MeterReading = 14  # bit[9] == 6 -> j(BasicMeterReading) else c (MeterReading)
    CurrentLevel = 26
    BasicCommand = 33  # q BasicCommand
    VariationCommand = 34  # d VariationCommand
    ScenarioCommand = 35  # k Scenario
    Variation = 37  # e Variation
    CurrentTransformerDefinition = 252  # l Current Transformer


#
# Functioning Mode / Heating Level
#

class FunctioningMode(IntEnum):
    Reduced = 0
    Moderato = 1
    Medio = 2
    Confort = 3
    Stop = 4
    AntiFrost = 5
    Special = 6
    Auto = 7
    Centralized = 8


FunctioningLevelMessage = Struct(
    "flags" / Bitwise(Struct(
        "manual" / Default(Flag, False),
        "duration" / Default(Flag, False),
        "f5" / Default(Flag, False),
        "f4" / Default(Flag, False),
        "mode" / Enum(BitsInteger(4), FunctioningMode)
    )),
    "duration" / Optional(Int16ub)
)

HeatingLevelMessage = FunctioningLevelMessage


#
# Temperature
#

class TemperatureAdapter(Adapter):
    def _decode(self, obj, context, path):
        return obj / 512

    def _encode(self, obj, context, path):
        return int(obj * 512)


TemperatureMessage = Struct(
    "temperature" / TemperatureAdapter(Int16ul)
)


#
# Basic Command
#

class BasicCommand(IntEnum):
    Off = 0
    On = 1
    Toggle = 2


BasicCommandMessage = Struct(
    "command" / Enum(Int8ub, BasicCommand)
)


#
# Variation
#

class VariationValueAdapter(Adapter):
    def _decode(self, obj, context, path):
        return obj & 0x7F

    def _encode(self, obj, context, path):
        return obj & 0x7F


VariationMessage = Struct(
    "value" / VariationValueAdapter(Int8ub)
)


#
# Variation Command
#

class VariationCommand(IntEnum):
    More = 1
    ShortReleasedMore = 129
    LongReleasedMore = 65
    Less = 2
    ShortReleasedLess = 130
    LongReleasedLess = 66
    Stop = 4


VariationCommandMessage = Struct(
    "command" / Enum(Int8ub, VariationCommand),
    "dummy1" / Default(Int8ub, 0),
    "dummy2" / Default(Int8ub, 0),
)

#
# Enrollment
#

EnrollmentMessage = Struct(
)


#
# Meter Reading Message
#


class ElectricityTariff(IntEnum):
    Base = 0
    EjpOffPeakDay = 32
    EjpPeakDay = 33
    DoubleOffPeakHour = 64
    DoublePeakHour = 65
    TempoBlueDayOffPeakHour = 96
    TempoBlueDayPeakHour = 97
    TempoWhiteDayOffPeakHour = 98
    TempoWhiteDayPeakHour = 99
    TempoRedDayOffPeakHour = 100
    TempoRedDayPeakHour = 101


class RegisterSelection(IntEnum):
    CurrentTransformer1 = 0
    CurrentTransformer2 = 1
    CurrentTransformer3 = 2
    Heating = 3
    HotWater = 4
    Total = 5
    Cooling = 7
    HeatingAndCooling = 8
    ElectricityProduction = 9


MeterReadingMessage = Struct(
    "selection" / Enum(Int8ub, RegisterSelection),
    "currentTariff" / Bitwise(Struct(
        "f7" / Default(Flag, False),
        "double" / Default(Flag, False),
        "ejp" / Default(Flag, False),
        "euro" / Default(Flag, False),
        "f3" / Default(Flag, False),
        "f2" / Default(Flag, False),
        "f1" / Default(Flag, False),
        "f0" / Default(Flag, False),
    )),
    "register" / BytesInteger(3, swapped=True)
)


#
# Functions
#

def x2d_crc(data):
    crc = 0
    for i in data:
        crc += int(i)
    return int((~crc) + 1) & 0xFFFF


class OffsettedEnd(Subconstruct):
    r"""
    Parses all bytes in the stream till EOF plus endoffset is reached.

    This is useful when GreedyBytes (or an other greedy construct) is followed by a fixed-size footer.

    Parsing determines the length of the stream and reads all bytes till EOF plus `endoffset` is reached, then defers to subcon using new BytesIO with said bytes. Building defers to subcon as-is. Size is undefined.

    :param endoffset: integer or context lambda, only negative offsets or 0 are allowed.
    :param subcon: Construct instance

    :raises StreamError: could not read enough bytes
    :raises StreamError: reads behind the stream (if endoffset is positive)

    Example::

        >>> d = Struct("data"/OffsettedEnd(-2, GreedyBytes), "footer"/Bytes(2))
        >>> d.parse(b"\x01\x02\x03\x04\x05")
        Container(data=b'\x01\x02\x03', footer=b'\x04\x05')
        >>> d.build(Container(data=b"\x01\x02\x03", footer=b"\x04\x05"))
        b'\x01\x02\x03\x04\x05'
    """

    def __init__(self, endoffset, subcon):
        super().__init__(subcon)
        self.endoffset = endoffset

    def _parse(self, stream, context, path):
        endoffset = construct.core.evaluate(self.endoffset, context)
        curpos = stream_tell(stream, path)
        stream_seek(stream, 0, 2, path)
        endpos = stream_tell(stream, path)
        stream_seek(stream, curpos, 0, path)
        length = endpos + endoffset - curpos
        data = stream_read(stream, length, path)
        if self.subcon is GreedyBytes:
            return data
        if type(self.subcon) is GreedyString:
            return data.decode(self.subcon.encoding)
        return self.subcon._parsereport(io.BytesIO(data), context, path)

    def _build(self, obj, stream, context, path):
        return self.subcon._build(obj, stream, context, path)


def _offset(context):
    return -2 if context.control.rolling_code else -0

_x2d_struct = Struct(
    "body" / OffsettedEnd(-2, RawCopy(Struct(
        "house" / Int16ub,
        "source" / Bitwise(Struct(
            "id" / BitsInteger(2),
            "type" / Enum(BitsInteger(6), Device)
        )),
        "recipient" / Bitwise(Struct(
            "f7" / Default(Flag, False),
            "f6" / Default(Flag, False),
            "f5" / Default(Flag, False),
            "f4" / Default(Flag, False),
            "zone" / BitsInteger(4)
        )),
        "transmitter" / Bitwise(Struct(
            "enrollment_requested" / Default(Flag, False),
            "internal_fault_detected" / Default(Flag, False),
            "box_opened" / Default(Flag, False),
            "battery_failing" / Default(Flag, False),
            "attribute" / Enum(BitsInteger(4), Attribute)
        )),
        "control" / Bitwise(Struct(
            "f7" / Default(Flag, False),
            "f6" / Default(Flag, False),
            "f5" / Default(Flag, False),
            "f4" / Default(Flag, False),
            "rolling_code" / Default(Flag, False),
            "answer_request" / Default(Flag, False),
            "f1" / Default(Flag, False),
            "f0" / Default(Flag, False)
        )),
        "data" / OffsettedEnd(_offset, RawCopy(Switch(lambda this: int(this.transmitter.attribute), {
            Attribute.WithData: Struct(
                "type" / Enum(Int8ub, MessageDataType),
                "content" / Optional(Switch(lambda this: int(this.type), {
                    MessageDataType.Enrollment: EnrollmentMessage,
                    MessageDataType.BasicCommand: BasicCommandMessage,
                    MessageDataType.HeatingLevel: HeatingLevelMessage,
                    MessageDataType.FunctioningLevel: FunctioningLevelMessage,
                    MessageDataType.VariationCommand: VariationCommandMessage,
                    MessageDataType.InternalTemperature: TemperatureMessage,
                    MessageDataType.MeterReading: MeterReadingMessage,
                    MessageDataType.CurrentLevel: HeatingLevelMessage,
                }, default=GreedyBytes))
            )
        }))),
        "rollingCode" / If(this.control.rolling_code, Int16ub),
    ))),
    "checksum" / Checksum(Int16ub, x2d_crc, this.body.data)
)


def parse_x2d_message(data):
    return _x2d_struct.parse(bytearray(data)).body.value


def format_x2d_message(msg):
    return _x2d_struct.build(dict(body=dict(value=msg)))

File: esphome/components/espx2d/x2d.c
#include "x2d.h"

#if defined(ARDUINO)
extern int arduino_printf(const char *__restrict, ...);
#define PRINT_FCT(...) arduino_printf(__VA_ARGS__)
#else
#include <stdio.h>
#define PRINT_FCT(...) printf(__VA_ARGS__)
#endif

#define VNS(x, y)   \
    {               \
        x##_##y, #y \
    }
#define ARRAY_SIZE(T) (sizeof(T) / sizeof(T[0]))

static struct
{
    X2D_DEVICE device;
    const char *name;
} device_str[] = {
    VNS(X2D_DEVICE, Tyxia_ZZAA),
    VNS(X2D_DEVICE, Calybox),
    VNS(X2D_DEVICE, Deltia_Emitter),
    VNS(X2D_DEVICE, Tydom_Panel_Controller),
    VNS(X2D_DEVICE, Tyxia_XXYY),
    VNS(X2D_DEVICE, Two_Buttons_Alarm_Remote),
    VNS(X2D_DEVICE, Four_Buttons_Alarm_Remote),
    VNS(X2D_DEVICE, Volumetric),
    VNS(X2D_DEVICE, Volumetric_Infrared),
    VNS(X2D_DEVICE, Volumetric_Pressure),
    VNS(X2D_DEVICE, Volumetric_Dual),
    VNS(X2D_DEVICE, Perimetric),
    VNS(X2D_DEVICE, Perimetric_Contact),
    VNS(X2D_DEVICE, Perimetric_Glass_Breakage),
    VNS(X2D_DEVICE, Perimetric_Sound),
    VNS(X2D_DEVICE, Perimetric_Roller_Shutter),
    VNS(X2D_DEVICE, Technical),
    VNS(X2D_DEVICE, Technical_Water),
    VNS(X2D_DEVICE, Technical_Gas),
    VNS(X2D_DEVICE, Technical_Fire),
    VNS(X2D_DEVICE, Technical_Smoke),
    VNS(X2D_DEVICE, Technical_Frost),
    VNS(X2D_DEVICE, Technical_Power_Outage),
    VNS(X2D_DEVICE, Technical_Phone_Outage),
    VNS(X2D_DEVICE, Technical_Freezer_Outage),

    VNS(X2D_DEVICE, USB_Key),
};

static const char *get_device_str(X2D_DEVICE device)
{
    for (int i = 0; i < ARRAY_SIZE(device_str); ++i)
    {
        if (device_str[i].device == device)
        {
            return device_str[i].name;
        }
    }
    return "(NULL)";
}

static struct
{
    X2D_ATTRIBUTE attribute;
    const char *name;
} attribute_str[] = {
    VNS(X2D_ATTRIBUTE, Simple),
    VNS(X2D_ATTRIBUTE, WithData),
};

static const char *get_attribute_str(X2D_ATTRIBUTE attribute)
{
    for (int i = 0; i < ARRAY_SIZE(attribute_str); ++i)
    {
        if (attribute_str[i].attribute == attribute)
        {
            return attribute_str[i].name;
        }
    }
    return "(NULL)";
}

static struct
{
    X2D_MESSAGE_DATA_TYPE message_data_type;
    const char *name;
} message_data_type_str[] = {
    VNS(X2D_MESSAGE_DATA_TYPE, Enrollment),
    VNS(X2D_MESSAGE_DATA_TYPE, HeatingLevel),
    VNS(X2D_MESSAGE_DATA_TYPE, FunctioningLevel),
    VNS(X2D_MESSAGE_DATA_TYPE, InternalTemperature),
    VNS(X2D_MESSAGE_DATA_TYPE, ExternalTemperature),
    VNS(X2D_MESSAGE_DATA_TYPE, MeterReading),
    VNS(X2D_MESSAGE_DATA_TYPE, CurrentLevel),
    VNS(X2D_MESSAGE_DATA_TYPE, BasicCommand),
    VNS(X2D_MESSAGE_DATA_TYPE, VariationCommand),
    VNS(X2D_MESSAGE_DATA_TYPE, ScenarioCommand),
    VNS(X2D_MESSAGE_DATA_TYPE, Variation),
    VNS(X2D_MESSAGE_DATA_TYPE, CurrentTransformerDefinition),
};

static const char *get_message_data_type_str(X2D_MESSAGE_DATA_TYPE message_data_type)
{
    for (int i = 0; i < ARRAY_SIZE(message_data_type_str); ++i)
    {
        if (message_data_type_str[i].message_data_type == message_data_type)
        {
            return message_data_type_str[i].name;
        }
    }
    return "(NULL)";
}

static struct
{
    X2D_FUNCTIONING_MODE functioning_mode;
    const char *name;
} functioning_mode_str[] = {
    VNS(X2D_FUNCTIONING_MODE, Reduced),
    VNS(X2D_FUNCTIONING_MODE, Moderato),
    VNS(X2D_FUNCTIONING_MODE, Medio),
    VNS(X2D_FUNCTIONING_MODE, Confort),
    VNS(X2D_FUNCTIONING_MODE, Stop),
    VNS(X2D_FUNCTIONING_MODE, AntiFrost),
    VNS(X2D_FUNCTIONING_MODE, Special),
    VNS(X2D_FUNCTIONING_MODE, Auto),
    VNS(X2D_FUNCTIONING_MODE, Centralized),
};

static const char *get_functioning_mode_str(X2D_FUNCTIONING_MODE functioning_mode)
{
    for (int i = 0; i < ARRAY_SIZE(functioning_mode_str); ++i)
    {
        if (functioning_mode_str[i].functioning_mode == functioning_mode)
        {
            return functioning_mode_str[i].name;
        }
    }
    return "(NULL)";
}

/*

TYPEDEF_ENUM_BEG(X2D_ELECTRICITY_TARIFF)
    X2D_ELECTRICITY_TARIFF_Base = 0,
    X2D_ELECTRICITY_TARIFF_EjpOffPeakDay = 32,
    X2D_ELECTRICITY_TARIFF_EjpPeakDay = 33,
    X2D_ELECTRICITY_TARIFF_DoubleOffPeakHour = 64,
    X2D_ELECTRICITY_TARIFF_DoublePeakHour = 65,
    X2D_ELECTRICITY_TARIFF_TempoBlueDayOffPeakHour = 96,
    X2D_ELECTRICITY_TARIFF_TempoBlueDayPeakHour = 97,
    X2D_ELECTRICITY_TARIFF_TempoWhiteDayOffPeakHour = 98,
    X2D_ELECTRICITY_TARIFF_TempoWhiteDayPeakHour = 99,
    X2D_ELECTRICITY_TARIFF_TempoRedDayOffPeakHour = 100,
    X2D_ELECTRICITY_TARIFF_TempoRedDayPeakHour = 101,
TYPEDEF_ENUM_END(X2D_ELECTRICITY_TARIFF)


TYPEDEF_ENUM_BEG(X2D_REGISTER_SELECTION)
    X2D_REGISTER_SELECTION_CurrentTransformer1 = 0,
    X2D_REGISTER_SELECTION_CurrentTransformer2 = 1,
    X2D_REGISTER_SELECTION_CurrentTransformer3 = 2,
    X2D_REGISTER_SELECTION_Heating = 3,
    X2D_REGISTER_SELECTION_HotWater = 4,
    X2D_REGISTER_SELECTION_Total = 5,
    X2D_REGISTER_SELECTION_Cooling = 7,
    X2D_REGISTER_SELECTION_HeatingAndCooling = 8,
    X2D_REGISTER_SELECTION_ElectricityProduction = 9,
TYPEDEF_ENUM_END(X2D_REGISTER_SELECTION)

*/

static struct
{
    X2D_ELECTRICITY_TARIFF electricity_tariff;
    const char *name;
} electricity_tariff_str[] = {
    VNS(X2D_ELECTRICITY_TARIFF, Base),
    VNS(X2D_ELECTRICITY_TARIFF, EjpOffPeakDay),
    VNS(X2D_ELECTRICITY_TARIFF, EjpPeakDay),
    VNS(X2D_ELECTRICITY_TARIFF, DoubleOffPeakHour),
    VNS(X2D_ELECTRICITY_TARIFF, DoublePeakHour),
    VNS(X2D_ELECTRICITY_TARIFF, TempoBlueDayOffPeakHour),
    VNS(X2D_ELECTRICITY_TARIFF, TempoBlueDayPeakHour),
    VNS(X2D_ELECTRICITY_TARIFF, TempoWhiteDayOffPeakHour),
    VNS(X2D_ELECTRICITY_TARIFF, TempoWhiteDayPeakHour),
    VNS(X2D_ELECTRICITY_TARIFF, TempoRedDayOffPeakHour),
    VNS(X2D_ELECTRICITY_TARIFF, TempoRedDayPeakHour),
};

static const char *get_electricity_tariff_str(X2D_ELECTRICITY_TARIFF electricity_tariff)
{
    for (int i = 0; i < ARRAY_SIZE(electricity_tariff_str); ++i)
    {
        if (electricity_tariff_str[i].electricity_tariff == electricity_tariff)
        {
            return electricity_tariff_str[i].name;
        }
    }
    return "(NULL)";
}

static struct
{
    X2D_REGISTER_SELECTION register_selection;
    const char *name;
} register_selection_str[] = {
    VNS(X2D_REGISTER_SELECTION, CurrentTransformer1),
    VNS(X2D_REGISTER_SELECTION, CurrentTransformer2),
    VNS(X2D_REGISTER_SELECTION, CurrentTransformer3),
    VNS(X2D_REGISTER_SELECTION, Heating),
    VNS(X2D_REGISTER_SELECTION, HotWater),
    VNS(X2D_REGISTER_SELECTION, Total),
    VNS(X2D_REGISTER_SELECTION, Cooling),
    VNS(X2D_REGISTER_SELECTION, HeatingAndCooling),
    VNS(X2D_REGISTER_SELECTION, ElectricityProduction),
};

static const char *get_register_selection_str(X2D_REGISTER_SELECTION register_selection)
{
    for (int i = 0; i < ARRAY_SIZE(register_selection_str); ++i)
    {
        if (register_selection_str[i].register_selection == register_selection)
        {
            return register_selection_str[i].name;
        }
    }
    return "(NULL)";
}

static const char *get_bool_str(bool b)
{
    return b ? "true" : "false";
}

static void print_data(uint8_t *data, size_t size)
{
    if (size < 1)
    {
        return;
    }

    X2D_MESSAGE_DATA_TYPE data_type = (X2D_MESSAGE_DATA_TYPE)data[0];
    PRINT_FCT("Data|Type: %s\n", get_message_data_type_str(data_type));
    size -= 1;
    data += 1;
    switch (data_type)
    {
    case X2D_MESSAGE_DATA_TYPE_Enrollment:
    {
        break;
    }
    case X2D_MESSAGE_DATA_TYPE_InternalTemperature:
    {
        if (size < sizeof(X2D_TEMPERATURE_MESSAGE))
        {
            return;
        }

        X2D_TEMPERATURE_MESSAGE *message = (X2D_TEMPERATURE_MESSAGE *)data;
        float temperature = (float)LU16_TO_UNS16(message->temperature) / 512.0f;
        float s_temperature = (temperature - (float)(int)temperature) * 10;
        PRINT_FCT("InternalTemperature|Temperature: %d.%d°\n", (int)temperature, (int)s_temperature);
        size -= sizeof(X2D_TEMPERATURE_MESSAGE);
        data += sizeof(X2D_TEMPERATURE_MESSAGE);
        break;
    }
    case X2D_MESSAGE_DATA_TYPE_ExternalTemperature:
    {
        if (size < sizeof(X2D_TEMPERATURE_MESSAGE))
        {
            return;
        }
        X2D_TEMPERATURE_MESSAGE *message = (X2D_TEMPERATURE_MESSAGE *)data;
        float temperature = (float)LU16_TO_UNS16(message->temperature) / 512.0f;
        float s_temperature = (temperature - (float)(int)temperature) * 10;
        PRINT_FCT("ExternalTemperature|Temperature: %d.%d°\n", (int)temperature, (int)s_temperature);
        size -= sizeof(X2D_TEMPERATURE_MESSAGE);
        data += sizeof(X2D_TEMPERATURE_MESSAGE);
        break;
    }
    case X2D_MESSAGE_DATA_TYPE_FunctioningLevel:
    {
        if (size < 1)
        {
            return;
        }
        X2D_FUNCTIONING_LEVEL_MESSAGE *message = (X2D_FUNCTIONING_LEVEL_MESSAGE *)data;
        PRINT_FCT("FunctioningLevel|FunctioningMode: %s\n", get_functioning_mode_str((X2D_FUNCTIONING_MODE)BITFIELD_GET_RANGE(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_MODE_3, X2D_FUNCTIONING_LEVEL_FLAG_MODE_0)));
        PRINT_FCT("FunctioningLevel|Manual: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_MANUAL)));
        if (BITFIELD_BIT_IS_SET(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_DURATION))
        {
            PRINT_FCT("FunctioningLevel|Duration %u\n", BU16_TO_UNS16(message->duration));
            size -= sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE);
            data += sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE);
        }
        else
        {
            size -= 1;
            data += 1;
        }

        break;
    }
    case X2D_MESSAGE_DATA_TYPE_HeatingLevel:
    {
        if (size < 1)
        {
            return;
        }
        X2D_FUNCTIONING_LEVEL_MESSAGE *message = (X2D_FUNCTIONING_LEVEL_MESSAGE *)data;
        PRINT_FCT("HeatingLevel|FunctioningMode: %s\n", get_functioning_mode_str((X2D_FUNCTIONING_MODE)BITFIELD_GET_RANGE(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_MODE_3, X2D_FUNCTIONING_LEVEL_FLAG_MODE_0)));
        PRINT_FCT("HeatingLevel|Manual: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_MANUAL)));
        if (size >= sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE) && BITFIELD_BIT_IS_SET(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_DURATION))
        {
            PRINT_FCT("HeatingLevel|Duration %u\n", BU16_TO_UNS16(message->duration));
            size -= sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE);
            data += sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE);
        }
        else
        {
            size -= 1;
            data += 1;
        }

        break;
    }
    case X2D_MESSAGE_DATA_TYPE_CurrentLevel:
    {
        if (size < 1)
        {
            return;
        }
        X2D_FUNCTIONING_LEVEL_MESSAGE *message = (X2D_FUNCTIONING_LEVEL_MESSAGE *)data;
        PRINT_FCT("CurrentLevel|FunctioningMode: %s\n", get_functioning_mode_str((X2D_FUNCTIONING_MODE)BITFIELD_GET_RANGE(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_MODE_3, X2D_FUNCTIONING_LEVEL_FLAG_MODE_0)));
        PRINT_FCT("CurrentLevel|Manual: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_MANUAL)));
        if (size >= sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE) && BITFIELD_BIT_IS_SET(X2D_FUNCTIONING_LEVEL_FLAG, message->flag, X2D_FUNCTIONING_LEVEL_FLAG_DURATION))
        {
            PRINT_FCT("CurrentLevel|Duration %u\n", BU16_TO_UNS16(message->duration));
            size -= sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE);
            data += sizeof(X2D_FUNCTIONING_LEVEL_MESSAGE);
        }
        else
        {
            size -= 1;
            data += 1;
        }

        break;
    }
    case X2D_MESSAGE_DATA_TYPE_MeterReading:
    {
        if (size < sizeof(X2D_METER_READING_MESSAGE))
        {
            return;
        }
        X2D_METER_READING_MESSAGE *message = (X2D_METER_READING_MESSAGE *)data;
        BITFIELD_VAR(X2D_CURRENT_TARIFF, flag);
        flag = message->flag;
        bool euro = BITFIELD_BIT_IS_SET(X2D_CURRENT_TARIFF, flag, X2D_CURRENT_TARIFF_EURO);
        BITFIELD_BIT_CLR(X2D_CURRENT_TARIFF, flag, X2D_CURRENT_TARIFF_EURO);
        uint32_t regValue = LU8_TO_UNS8(message->registerHigh) << 16 | LU16_TO_UNS16(message->registerLow);

        PRINT_FCT("MeterReading|Selection: %s\n", get_functioning_mode_str(message->selection));
        PRINT_FCT("MeterReading|Tariff: %s\n", get_electricity_tariff_str((X2D_ELECTRICITY_TARIFF)BITFIELD_GET(X2D_CURRENT_TARIFF, flag)));
        PRINT_FCT("MeterReading|Register: %u\n", regValue);
        PRINT_FCT("MeterReading|Unit: %s\n", euro ? "Euro" : "Kwh");
        size -= sizeof(X2D_METER_READING_MESSAGE);
        data += sizeof(X2D_METER_READING_MESSAGE);
        break;
    }
    }
}

void x2d_print(uint8_t *data, size_t size)
{
    if (size < sizeof(X2D_BODY) + sizeof(X2D_FOOTER))
    {
        PRINT_FCT("Invalid X2D data size\n");
        return;
    }

    size -= sizeof(X2D_FOOTER);
    X2D_FOOTER *footer = (X2D_FOOTER *)&data[size];
    uint16_t checksum = x2d_compute_checksum(data, size);
    if (checksum != BU16_TO_UNS16(footer->checksum))
    {
        PRINT_FCT("Invalid checksum\n");
        return;
    }

    X2D_BODY *body = (X2D_BODY *)data;
    PRINT_FCT("House: %u\n", BU16_TO_UNS16(body->house));
    PRINT_FCT("Source|Id: %u\n", BITFIELD_GET_RANGE(X2D_SOURCE, body->source, X2D_SOURCE_ID_1, X2D_SOURCE_ID_0));
    PRINT_FCT("Source|Type: %s\n", get_device_str((X2D_DEVICE)BITFIELD_GET_RANGE(X2D_SOURCE, body->source, X2D_SOURCE_TYPE_5, X2D_SOURCE_TYPE_0)));
    PRINT_FCT("Recipient|Zone: %u\n", BITFIELD_GET_RANGE(X2D_RECIPIENT, body->recipient, X2D_RECIPIENT_ZONE_3, X2D_RECIPIENT_ZONE_0));
    PRINT_FCT("Transmitter|Attribute: %s\n", get_attribute_str((X2D_ATTRIBUTE)BITFIELD_GET_RANGE(X2D_TRANSMITER, body->transmitter, X2D_TRANSMITER_ATTRIBUTE_3, X2D_TRANSMITER_ATTRIBUTE_0)));
    PRINT_FCT("Transmitter|BatteryFailing: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_TRANSMITER, body->transmitter, X2D_TRANSMITER_BATTERY_FAILLING)));
    PRINT_FCT("Transmitter|BoxOpened: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_TRANSMITER, body->transmitter, X2D_TRANSMITER_BOX_OPENED)));
    PRINT_FCT("Transmitter|InternalFaultDetected: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_TRANSMITER, body->transmitter, X2D_TRANSMITER_INTERNAL_FAULT_DETECTED)));
    PRINT_FCT("Transmitter|EnrollmentRequested: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_TRANSMITER, body->transmitter, X2D_TRANSMITER_ENROLLMENT_REQUESTED)));
    PRINT_FCT("Control|AnswerRequested: %s\n", get_bool_str(BITFIELD_BIT_IS_SET(X2D_CONTROL, body->control, X2D_CONTROL_ANSWER_REQUESTED)));

    size -= sizeof(X2D_BODY);
    data += sizeof(X2D_BODY);

    if (BITFIELD_BIT_IS_SET(X2D_CONTROL, body->control, X2D_CONTROL_ROLLING_CODE))
    {
        size -= sizeof(X2D_ROLLING_CODE_FOOTER);
        X2D_ROLLING_CODE_FOOTER *message = (X2D_ROLLING_CODE_FOOTER *)&data[size];
        PRINT_FCT("RollingCode: %u\n", BU16_TO_UNS16(message->rollingCode));
    }

    print_data(data, size);
}

size_t x2d_get_frame_size(uint8_t *data, size_t size)
{
    if (size < sizeof(X2D_BODY) + sizeof(X2D_FOOTER))
    {
        return 0;
    }

    size -= sizeof(X2D_FOOTER);
    X2D_FOOTER *footer = (X2D_FOOTER *)&data[size];
    uint16_t checksum = x2d_compute_checksum(data, size);
    if (checksum != BU16_TO_UNS16(footer->checksum))
    {
        return 0;
    }
    return size;
}

uint16_t x2d_compute_checksum(uint8_t *data, size_t size)
{
    uint16_t checksum = 0;
    for (int i = 0; i < size; ++i)
    {
        checksum += data[i];
    }
    return ~checksum + 1;
}